File: scoreboard/tui.py
"""Live TUI status board using Rich."""

import json
import time
from dataclasses import dataclass
from datetime import datetime, timezone

from rich import box
from rich.console import Console, Group
from rich.live import Live
from rich.panel import Panel
from rich.table import Table
from rich.text import Text

from .verify import StatusReport, verify_report, parse_report

# Dreadnode color palette
C_SUCCESS = "#68c147"
C_ERROR = "#e44f4f"
C_WARNING = "#c8ac4a"
C_INFO = "#4689bf"
C_BRAND = "#ca5e44"
C_ACCENT = "#ef562f"
C_PURPLE = "#a650fb"
C_TEAL = "#20dfc8"
C_FG = "#e2e7ec"
C_FG_SUBTLE = "#c1c6cc"
C_FG_MUTED = "#9da0a5"
C_FG_FAINTEST = "#686d73"
C_BORDER = "#2b343f"

# Group display config
GROUP_CONFIG = {
    "credentials": {
        "title": "CREDENTIALS DISCOVERED",
        "short": "CREDENTIALS",
        "color": f"bold {C_BRAND}",
    },
    "hosts": {
        "title": "HOSTS COMPROMISED",
        "short": "HOSTS",
        "color": f"bold {C_BRAND}",
    },
    "domains": {
        "title": "DOMAINS OWNED",
        "short": "DOMAINS",
        "color": f"bold {C_BRAND}",
    },
    "techniques": {
        "title": "ATTACK TECHNIQUES USED",
        "short": "ATTACK TECHNIQUES",
        "color": f"bold {C_BRAND}",
    },
}

# Layout: left column groups, right column groups
LEFT_GROUPS = ["domains", "hosts", "techniques"]
RIGHT_GROUPS = ["credentials"]


@dataclass
class PollState:
    """Tracks polling status for the footer bar."""

    last_poll_time: float = 0.0
    poll_interval: float = 3.0
    last_result: str = "waiting"  # "ok", "no_file", "error", "waiting"
    last_error: str = ""
    finding_count: int = 0
    report_path: str = "/tmp/report.jsonl"


def build_header(status: StatusReport, agent_id: str, elapsed: str) -> Table:
    """Build the header bar with colorful stats."""
    table = Table(show_header=False, show_edge=False, pad_edge=False, expand=True)
    table.add_column(ratio=1)
    table.add_column(ratio=1, justify="right")

    summary = Text()
    first = True
    for group, stats in status.groups.items():
        cfg = GROUP_CONFIG.get(group, {"title": group.upper(), "color": "white"})
        label = cfg.get("short", cfg["title"])
        color = cfg["color"]

        if not first:
            summary.append("  |  ", style=C_FG_FAINTEST)
        summary.append(f"{label} ", style=color)
        achieved = stats["achieved"]
        total = stats["total"]
        summary.append(f"{achieved}", style=f"bold {C_SUCCESS}")
        summary.append("/", style=C_FG)
        summary.append(f"{total}", style=C_INFO)
        first = False

    table.add_row(summary, Text(f"Agent: {agent_id}  |  {elapsed}", style=C_FG_MUTED))
    return table


def build_group_section(
    group: str, stats: dict, verified: list, answer_key: dict
) -> Table:
    """Build a section for one milestone group."""
    cfg = GROUP_CONFIG.get(group, {"title": group.upper(), "color": "bold white"})
    achieved = stats["achieved"]
    total = stats["total"]

    table = Table(
        show_header=False,
        show_edge=False,
        pad_edge=True,
        title=f"  {cfg['title']}  ({achieved}/{total})",
        title_style=cfg["color"],
        title_justify="left",
        expand=True,
        box=box.SIMPLE,
        padding=(0, 1, 0, 0),
    )
    table.add_column("status", width=4, no_wrap=True)
    table.add_column("label", ratio=1)
    table.add_column("time", width=10, justify="right", no_wrap=True)

    achieved_ids = {}
    for vo in verified:
        if vo.group == group and vo.verified:
            achieved_ids[vo.objective_id] = vo

    group_objectives = [
        o for o in answer_key.get("objectives", []) if o["group"] == group
    ]

    for obj in group_objectives:
        vo = achieved_ids.get(obj["id"])
        if vo:
            ts = _format_ts(vo.timestamp)
            table.add_row(
                Text("[x]", style=f"bold {C_SUCCESS}"),
                Text(obj["label"]),
                Text(ts, style=C_FG_MUTED),
            )
        else:
            hint = obj.get("hint", "") or ""
            label_text = obj["label"]
            if hint:
                label_text += f"  ({hint})"
            table.add_row(
                Text("[ ]", style=C_FG_FAINTEST),
                Text(label_text, style=C_FG_FAINTEST),
                Text(""),
            )

    return table


def _format_ts(timestamp: str) -> str:
    if not timestamp:
        return ""
    try:
        dt = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
        return dt.strftime("%H:%M:%S")
    except ValueError:
        return timestamp[:8]


def build_poll_footer(poll: PollState) -> Text:
    """Build the polling status footer line."""
    now = time.monotonic()
    since_poll = now - poll.last_poll_time
    next_in = max(0, poll.poll_interval - since_poll)

    footer = Text()

    # Status indicator
    if poll.last_result == "ok":
        footer.append("  CONNECTED", style=f"bold {C_SUCCESS}")
        footer.append(f"  ({poll.finding_count} findings)", style=C_FG_MUTED)
    elif poll.last_result == "no_file":
        footer.append("  WAITING FOR REPORT", style=f"bold {C_WARNING}")
        footer.append(f"  ({poll.report_path})", style=C_FG_FAINTEST)
    elif poll.last_result == "error":
        footer.append("  FETCH ERROR", style=f"bold {C_ERROR}")
        if poll.last_error:
            footer.append(f"  ({poll.last_error})", style=C_FG_MUTED)
    else:
        footer.append("  CONNECTING...", style=f"bold {C_INFO}")

    # Countdown
    footer.append(f"  |  next poll: {next_in:.0f}s", style=C_FG_FAINTEST)

    return footer


def build_status_board(
    status: StatusReport,
    agent_id: str,
    start_time: datetime | None,
    answer_key: dict,
    poll: PollState | None = None,
) -> Panel:
    """Build the full status board panel with two-column layout."""
    if start_time:
        elapsed = str(
            datetime.now(timezone.utc).replace(tzinfo=None) - start_time
        ).split(".")[0]
    else:
        elapsed = "--:--:--"

    header = build_header(status, agent_id, elapsed)

    # Build left column sections
    left_sections = []
    for group in LEFT_GROUPS:
        stats = status.groups.get(group)
        if not stats or stats["total"] == 0:
            continue
        left_sections.append(
            build_group_section(group, stats, status.verified, answer_key)
        )
        left_sections.append(Text(""))

    # Build right column sections
    right_sections = []
    for group in RIGHT_GROUPS:
        stats = status.groups.get(group)
        if not stats or stats["total"] == 0:
            continue
        right_sections.append(
            build_group_section(group, stats, status.verified, answer_key)
        )
        right_sections.append(Text(""))

    left_col = Group(*left_sections) if left_sections else Text("")
    right_col = Group(*right_sections) if right_sections else Text("")

    columns = Table(
        show_header=False,
        show_edge=False,
        pad_edge=False,
        expand=True,
        border_style=C_BORDER,
        show_lines=False,
    )
    columns.add_column(ratio=1, vertical="top")
    columns.add_column(ratio=1, vertical="top")
    columns.add_row(left_col, right_col)

    # Footer
    footer_parts = []
    if status.unmatched_findings:
        footer_parts.append(
            Text(
                f"  + {len(status.unmatched_findings)} additional finding(s) reported",
                style=f"italic {C_FG_FAINTEST}",
            )
        )
    if poll:
        footer_parts.append(build_poll_footer(poll))

    content = Group(header, Text(""), columns, *footer_parts)

    return Panel(
        content,
        title=f"[bold {C_BRAND}]DreadGOAD STATUS BOARD[/bold {C_BRAND}]",
        border_style=C_BRAND,
        expand=True,
    )


def run_tui(
    transport,
    answer_key: dict,
    poll_interval: float = 3.0,
    report_path: str = "/tmp/report.jsonl",
):
    """Main TUI loop. Polls transport for report updates and refreshes display."""
    console = Console()
    agent_id = "dreadnode-agent"
    start_time = None
    last_report_hash = None

    empty_report = {"agent_id": "dreadnode-agent", "findings": []}
    status = verify_report(empty_report, answer_key)
    poll = PollState(poll_interval=poll_interval, report_path=report_path)

    console.print(
        f"[bold {C_BRAND}]DreadGOAD Status Board[/bold {C_BRAND}] starting..."
    )
    console.print(f"Polling every {poll_interval}s. Press Ctrl+C to exit.\n")

    with Live(
        build_status_board(status, agent_id, start_time, answer_key, poll),
        console=console,
        refresh_per_second=2,
    ) as live:
        while True:
            try:
                # Poll for report
                try:
                    raw = transport.fetch_report()
                    poll.last_error = ""
                except Exception as e:
                    raw = None
                    poll.last_result = "error"
                    poll.last_error = str(e)
                poll.last_poll_time = time.monotonic()

                if raw:
                    poll.last_result = "ok"
                    poll.last_error = ""
                    report_hash = hash(raw)
                    if report_hash != last_report_hash:
                        last_report_hash = report_hash
                        report = parse_report(raw)
                        agent_id = report.get("agent_id", "dreadnode-agent")
                        poll.finding_count = len(report.get("findings", []))
                        if report.get("start_time") and not start_time:
                            try:
                                start_time = datetime.fromisoformat(
                                    report["start_time"].replace("Z", "+00:00")
                                ).replace(tzinfo=None)
                            except ValueError:
                                pass
                        status = verify_report(report, answer_key)
                elif poll.last_result != "error":
                    poll.last_result = "no_file"

                # Update display at higher rate for countdown
                for _ in range(int(poll_interval * 2)):
                    live.update(
                        build_status_board(
                            status, agent_id, start_time, answer_key, poll
                        )
                    )
                    time.sleep(0.5)

            except KeyboardInterrupt:
                break
            except json.JSONDecodeError:
                poll.last_result = "error"
                time.sleep(poll_interval)
                continue

    console.print(f"\n[bold {C_FG}]Final status:[/bold {C_FG}]")
    console.print(build_status_board(status, agent_id, start_time, answer_key, poll))

File: cli/internal/scoreboard/tui.go
package scoreboard

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// Dreadnode color palette.
const (
	cSuccess    = "#68c147"
	cError      = "#e44f4f"
	cWarning    = "#c8ac4a"
	cInfo       = "#4689bf"
	cBrand      = "#ca5e44"
	cFG         = "#e2e7ec"
	cFGMuted    = "#9da0a5"
	cFGFaintest = "#686d73"
)

var (
	styleTitle    = lipgloss.NewStyle().Foreground(lipgloss.Color(cBrand)).Bold(true)
	styleBorder   = lipgloss.NewStyle().Foreground(lipgloss.Color(cBrand))
	styleGroupHdr = lipgloss.NewStyle().Foreground(lipgloss.Color(cBrand)).Bold(true)
	styleAchieved = lipgloss.NewStyle().Foreground(lipgloss.Color(cSuccess)).Bold(true)
	styleTotal    = lipgloss.NewStyle().Foreground(lipgloss.Color(cInfo))
	styleSep      = lipgloss.NewStyle().Foreground(lipgloss.Color(cFGFaintest))
	styleMuted    = lipgloss.NewStyle().Foreground(lipgloss.Color(cFGMuted))
	styleFaint    = lipgloss.NewStyle().Foreground(lipgloss.Color(cFGFaintest))
	styleFG       = lipgloss.NewStyle().Foreground(lipgloss.Color(cFG))
	styleOK       = lipgloss.NewStyle().Foreground(lipgloss.Color(cSuccess)).Bold(true)
	styleWarn     = lipgloss.NewStyle().Foreground(lipgloss.Color(cWarning)).Bold(true)
	styleErr      = lipgloss.NewStyle().Foreground(lipgloss.Color(cError)).Bold(true)
	styleInfo     = lipgloss.NewStyle().Foreground(lipgloss.Color(cInfo)).Bold(true)
)

var groupTitles = map[string]string{
	"credentials": "CREDENTIALS DISCOVERED",
	"hosts":       "HOSTS COMPROMISED",
	"domains":     "DOMAINS OWNED",
	"techniques":  "ATTACK TECHNIQUES USED",
}

var groupShort = map[string]string{
	"credentials": "CREDENTIALS",
	"hosts":       "HOSTS",
	"domains":     "DOMAINS",
	"techniques":  "ATTACK TECHNIQUES",
}

var leftGroups = []string{"domains", "hosts", "techniques"}
var rightGroups = []string{"credentials"}

type pollResult int

const (
	pollWaiting pollResult = iota
	pollOK
	pollNoFile
	pollError
)

// TUIConfig configures the live status board.
type TUIConfig struct {
	Transport    Transport
	AnswerKey    *AnswerKey
	PollInterval time.Duration
	ReportPath   string // for display in the footer
}

// RunTUI starts the interactive status board. It returns when the user
// quits (q/ctrl-c) or the context is cancelled.
func RunTUI(ctx context.Context, cfg TUIConfig) error {
	if cfg.PollInterval <= 0 {
		cfg.PollInterval = 3 * time.Second
	}
	m := newModel(ctx, cfg)
	p := tea.NewProgram(m, tea.WithAltScreen(), tea.WithContext(ctx))
	_, err := p.Run()
	return err
}

// RenderStatic returns the status board as a single string (used by the demo
// command to print one snapshot without entering an alt-screen TUI).
func RenderStatic(status *StatusReport, ak *AnswerKey, agentID string, startTime time.Time) string {
	width := 120
	return renderBoard(status, ak, agentID, startTime, nil, width)
}

type model struct {
	ctx        context.Context
	cfg        TUIConfig
	status     *StatusReport
	report     *Report
	startTime  time.Time
	width      int
	height     int
	lastPollAt time.Time
	pollState  pollResult
	pollErr    string
	lastHash   uint64
	quitting   bool
}

func newModel(ctx context.Context, cfg TUIConfig) *model {
	empty := &Report{AgentID: "dreadnode-agent"}
	return &model{
		ctx:    ctx,
		cfg:    cfg,
		status: VerifyReport(empty, cfg.AnswerKey),
		report: empty,
	}
}

func (m *model) Init() tea.Cmd {
	return tea.Batch(m.pollCmd(), tickCmd())
}

type pollMsg struct {
	raw  string
	err  error
	when time.Time
}
type tickMsg struct{ t time.Time }

func (m *model) pollCmd() tea.Cmd {
	return func() tea.Msg {
		ctx, cancel := context.WithTimeout(m.ctx, 30*time.Second)
		defer cancel()
		raw, err := m.cfg.Transport.FetchReport(ctx)
		return pollMsg{raw: raw, err: err, when: time.Now()}
	}
}

func tickCmd() tea.Cmd {
	return tea.Tick(500*time.Millisecond, func(t time.Time) tea.Msg { return tickMsg{t} })
}

func (m *model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.WindowSizeMsg:
		m.width = msg.Width
		m.height = msg.Height
	case tea.KeyMsg:
		switch msg.String() {
		case "q", "ctrl+c", "esc":
			m.quitting = true
			return m, tea.Quit
		case "r":
			return m, m.pollCmd()
		}
	case pollMsg:
		m.lastPollAt = msg.when
		switch {
		case msg.err == nil:
			m.pollState = pollOK
			m.pollErr = ""
			h := simpleHash(msg.raw)
			if h != m.lastHash {
				m.lastHash = h
				m.report = ParseReport(msg.raw)
				if st, err := time.Parse(time.RFC3339, m.report.StartTime); err == nil && m.startTime.IsZero() {
					m.startTime = st
				}
				m.status = VerifyReport(m.report, m.cfg.AnswerKey)
			}
		case errors.Is(msg.err, ErrNoReport):
			m.pollState = pollNoFile
			m.pollErr = ""
		default:
			m.pollState = pollError
			m.pollErr = msg.err.Error()
		}
		// Schedule next poll
		next := tea.Tick(m.cfg.PollInterval, func(time.Time) tea.Msg {
			return pollKickMsg{}
		})
		return m, next
	case pollKickMsg:
		return m, m.pollCmd()
	case tickMsg:
		return m, tickCmd()
	}
	return m, nil
}

type pollKickMsg struct{}

func (m *model) View() string {
	if m.quitting {
		return ""
	}
	width := m.width
	if width <= 0 {
		width = 120
	}
	pollSnap := &pollSnapshot{
		state:        m.pollState,
		errMsg:       m.pollErr,
		findingCount: len(m.report.Findings),
		reportPath:   m.cfg.ReportPath,
		lastPollAt:   m.lastPollAt,
		interval:     m.cfg.PollInterval,
	}
	return renderBoard(m.status, m.cfg.AnswerKey, m.report.AgentID, m.startTime, pollSnap, width)
}

type pollSnapshot struct {
	state        pollResult
	errMsg       string
	findingCount int
	reportPath   string
	lastPollAt   time.Time
	interval     time.Duration
}

func renderBoard(status *StatusReport, ak *AnswerKey, agentID string, startTime time.Time, poll *pollSnapshot, width int) string {
	innerWidth := width - 4 // 2 chars border + 2 chars padding (1 each side)
	if innerWidth < 40 {
		innerWidth = 40
	}
	header := renderHeader(status, agentID, startTime, innerWidth)

	colWidth := (innerWidth - 2) / 2
	if colWidth < 30 {
		colWidth = 30
	}
	left := renderColumn(leftGroups, status, ak, colWidth)
	right := renderColumn(rightGroups, status, ak, colWidth)
	cols := lipgloss.JoinHorizontal(lipgloss.Top, left, "  ", right)

	parts := []string{header, "", cols}
	if len(status.UnmatchedFindings) > 0 {
		parts = append(parts, "",
			styleFaint.Italic(true).Render(fmt.Sprintf("  + %d additional finding(s) reported", len(status.UnmatchedFindings))))
	}
	if poll != nil {
		parts = append(parts, "", renderPollFooter(poll))
		parts = append(parts, styleFaint.Render("  q/ctrl-c quit · r reload"))
	}

	return panelWithTitle("DreadGOAD STATUS BOARD", strings.Join(parts, "\n"), width)
}

// panelWithTitle frames `body` in a rounded border with `title` embedded in
// the top edge.
func panelWithTitle(title, body string, width int) string {
	innerWidth := width - 4 // border (2) + padding (2)
	if innerWidth < 1 {
		innerWidth = 1
	}

	titleText := " " + title + " "
	titleVis := lipgloss.Width(titleText)
	leadDashes := 2
	trailDashes := innerWidth + 2 - leadDashes - titleVis
	if trailDashes < 1 {
		trailDashes = 1
	}
	top := styleBorder.Render("╭"+strings.Repeat("─", leadDashes)) +
		styleTitle.Render(titleText) +
		styleBorder.Render(strings.Repeat("─", trailDashes)+"╮")

	bottom := styleBorder.Render("╰" + strings.Repeat("─", innerWidth+2) + "╯")

	var rows []string
	rows = append(rows, top)
	for _, line := range strings.Split(body, "\n") {
		pad := innerWidth - lipgloss.Width(line)
		if pad < 0 {
			line = truncate(line, innerWidth)
			pad = 0
		}
		rows = append(rows, styleBorder.Render("│")+" "+line+strings.Repeat(" ", pad)+" "+styleBorder.Render("│"))
	}
	rows = append(rows, bottom)
	return strings.Join(rows, "\n")
}

func renderHeader(status *StatusReport, agentID string, startTime time.Time, width int) string {
	left := strings.Builder{}
	first := true
	groupOrder := []string{"credentials", "hosts", "domains", "techniques"}
	for _, g := range groupOrder {
		stats, ok := status.Groups[g]
		if !ok {
			continue
		}
		if !first {
			left.WriteString(styleSep.Render("  |  "))
		}
		first = false
		short := groupShort[g]
		if short == "" {
			short = strings.ToUpper(g)
		}
		left.WriteString(styleGroupHdr.Render(short + " "))
		left.WriteString(styleAchieved.Render(fmt.Sprintf("%d", stats.Achieved)))
		left.WriteString(styleFG.Render("/"))
		left.WriteString(styleTotal.Render(fmt.Sprintf("%d", stats.Total)))
	}

	elapsed := "--:--:--"
	if !startTime.IsZero() {
		elapsed = formatDuration(time.Since(startTime))
	}
	right := styleMuted.Render(fmt.Sprintf("Agent: %s  |  %s", agentID, elapsed))

	leftStr := left.String()
	pad := width - lipgloss.Width(leftStr) - lipgloss.Width(right)
	if pad < 1 {
		pad = 1
	}
	return leftStr + strings.Repeat(" ", pad) + right
}

func renderColumn(groups []string, status *StatusReport, ak *AnswerKey, width int) string {
	var sections []string
	for _, g := range groups {
		stats, ok := status.Groups[g]
		if !ok || stats.Total == 0 {
			continue
		}
		sections = append(sections, renderGroupSection(g, stats, status.Verified, ak, width))
	}
	return lipgloss.JoinVertical(lipgloss.Left, sections...)
}

func renderGroupSection(group string, stats *GroupStats, verified []VerifiedObjective, ak *AnswerKey, width int) string {
	title := groupTitles[group]
	if title == "" {
		title = strings.ToUpper(group)
	}
	hdr := styleGroupHdr.Render(fmt.Sprintf("  %s  (%d/%d)", title, stats.Achieved, stats.Total))

	achieved := map[string]VerifiedObjective{}
	for _, vo := range verified {
		if vo.Group == group && vo.Verified {
			achieved[vo.ObjectiveID] = vo
		}
	}

	rowWidth := width
	timeColWidth := 10
	statusColWidth := 4
	labelWidth := rowWidth - timeColWidth - statusColWidth - 2
	if labelWidth < 10 {
		labelWidth = 10
	}

	var rows []string
	for _, obj := range ak.Objectives {
		if obj.Group != group {
			continue
		}
		vo, ok := achieved[obj.ID]
		var statusCell, labelCell, timeCell string
		if ok {
			statusCell = styleOK.Render("[x] ")
			labelCell = styleFG.Render(truncate(obj.Label, labelWidth))
			timeCell = styleMuted.Render(formatTS(vo.Timestamp))
		} else {
			statusCell = styleFaint.Render("[ ] ")
			label := obj.Label
			if obj.Hint != "" {
				label = fmt.Sprintf("%s  (%s)", label, obj.Hint)
			}
			labelCell = styleFaint.Render(truncate(label, labelWidth))
			timeCell = ""
		}
		labelCell = padRight(labelCell, labelWidth)
		timeCell = padRight(timeCell, timeColWidth)
		rows = append(rows, statusCell+labelCell+timeCell)
	}
	return hdr + "\n" + strings.Join(rows, "\n") + "\n"
}

func renderPollFooter(p *pollSnapshot) string {
	since := time.Since(p.lastPollAt)
	if p.lastPollAt.IsZero() {
		since = 0
	}
	next := p.interval - since
	if next < 0 {
		next = 0
	}

	b := strings.Builder{}
	switch p.state {
	case pollOK:
		b.WriteString(styleOK.Render("  CONNECTED"))
		b.WriteString(styleMuted.Render(fmt.Sprintf("  (%d findings)", p.findingCount)))
	case pollNoFile:
		b.WriteString(styleWarn.Render("  WAITING FOR REPORT"))
		b.WriteString(styleFaint.Render(fmt.Sprintf("  (%s)", p.reportPath)))
	case pollError:
		b.WriteString(styleErr.Render("  FETCH ERROR"))
		if p.errMsg != "" {
			b.WriteString(styleMuted.Render(fmt.Sprintf("  (%s)", truncate(p.errMsg, 80))))
		}
	default:
		b.WriteString(styleInfo.Render("  CONNECTING..."))
	}
	b.WriteString(styleFaint.Render(fmt.Sprintf("  |  next poll: %ds", int(next.Seconds()))))
	return b.String()
}

func formatTS(ts string) string {
	if ts == "" {
		return ""
	}
	if t, err := time.Parse(time.RFC3339, ts); err == nil {
		return t.Format("15:04:05")
	}
	if len(ts) > 8 {
		return ts[:8]
	}
	return ts
}

func formatDuration(d time.Duration) string {
	if d < 0 {
		d = 0
	}
	h := int(d.Hours())
	m := int(d.Minutes()) % 60
	s := int(d.Seconds()) % 60
	return fmt.Sprintf("%d:%02d:%02d", h, m, s)
}

func padRight(s string, w int) string {
	pad := w - lipgloss.Width(s)
	if pad <= 0 {
		return s
	}
	return s + strings.Repeat(" ", pad)
}

func truncate(s string, w int) string {
	if w <= 0 {
		return ""
	}
	if lipgloss.Width(s) <= w {
		return s
	}
	if w <= 1 {
		return s[:1]
	}
	// naive byte-level truncation; lab labels are ASCII
	if w > len(s) {
		return s
	}
	return s[:w-1] + "…"
}

// simpleHash is a non-cryptographic hash used only to detect report changes.
func simpleHash(s string) uint64 {
	var h uint64 = 1469598103934665603
	for i := 0; i < len(s); i++ {
		h ^= uint64(s[i])
		h *= 1099511628211
	}
	return h
}
